### DISCUSSION TITLE
py-capellambse and sequence links in Functional Chain Description diagrams

### DISCUSSION THREAD

**Discussion by khammoutene** [dbinfrago/py-capellambse#173] [Q&A] [answered]
Hello,

From the code below, I having problems getting the "Sequence link" objects that exist in a Functional Chain Description diagram. Did I miss anything?

````
for` aCapability in model.sa.all_capabilities:
    print()
    print("------------------------------- System capability: " + aCapability.name + "-------------------------------")
    print()
    print(aCapability)
    for aFunctionalChain in aCapability.owned_chains:
        print()
        print("------------------------------------------- Owned Functional chain: " + aFunctionalChain.name + "-------------------------------")
        print()
        print(aFunctionalChain)
        for anInvolvedFunction in aFunctionalChain.involved_functions:
            print()
            print("------------------------------------------ involved function: ")
            print()
            print(anInvolvedFunction)
        for anInvolved in aFunctionalChain.involved:
            print()
            print("------------------------------------------ involved: ")
            print()
            print(anInvolved)
        for aConstraint in aFunctionalChain.constraints:
            print()
            print("------------------------------------------ constraint: ")
            print()
            print(aConstraint)
            print(aConstraint.specification)
        for aLink in aFunctionalChain.involved_links:
            print()
            print("------------------------------------------ link: ")
            print()
            print(aLink)``
```



thank you

**Comment by ewuerger** [comment, marked as answer]
Hi,

and sorry for the late reply. Whenever you deal with diagrams you can investigate its `nodes` attribute. E.g.:
![image](https://user-images.githubusercontent.com/50786483/200562810-1881db6a-053b-44d3-8cc4-820ab6c39193.png)

This is what I'm receiving for our test [SFCD] diagram:
![image](https://user-images.githubusercontent.com/50786483/200562920-c42cbaed-0c59-4c78-a2fa-4359f87410ff.png)

Now you can iterate over it and filter with the condition `.xtype.endswith("SequenceLink")` which is more reliable since instance checks require that an declaration in our model submodule exists.

This information can also be retrieved from the generated documentation, [here](https://dsd-dbs.github.io/py-capellambse/code/capellambse.model.html#capellambse.model.diagram.AbstractDiagram.nodes). But we want to tackle a mayor refactoring for the aird-parsing and resulting diagram export for a long time. This should then come with a nice notebook showing all the capabilities of our `Diagram` object and best practices.

Cheers.
